[Toolsets] Unable to get list of tools for notion mcp toolset

### Name and Version

ai-dial-core

### What steps will reproduce the bug?

Try to get tools for toolset with `https://mcp.notion.com/mcp` endpoint

### What is the expected behavior?

Tools are successfully retrieved

### What do you see instead?

```
{"Body":"Failed to fetch tools from MCP server for toolset: toolsets/public/27.03 notion toolset__0.0.3","Attributes":{"instant":"2026-05-21T08:52:00.736872999Z","loggerName":"com.epam.aidial.core.server.controller.ToolSetToolsController","threadName":"dial.x-virtual-thread-439248","user.project":null,"user.id":"26fa5438-5503-41a7-af7b-2dd7ad794c2c","request.method":"GET","request.uri":"/v1/toolset/toolsets/public/27.03%20notion%20toolset__0.0.3/tools","error":true,"exception.type":"java.lang.IllegalArgumentException","exception.message":"Cannot deserialize value of type `java.lang.Boolean` from Object value (token `JsonToken.START_OBJECT`)\n at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: io.modelcontextprotocol.spec.McpSchema$ListToolsResult[\"tools\"]->java.util.ArrayList[0]->io.modelcontextprotocol.spec.McpSchema$Tool[\"inputSchema\"]->io.modelcontextprotocol.spec.McpSchema$JsonSchema[\"additionalProperties\"])","exception.stacktrace":"java.lang.IllegalArgumentException: Cannot deserialize value of type `java.lang.Boolean` from Object value (token `JsonToken.START_OBJECT`)\n at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: io.modelcontextprotocol.spec.McpSchema$ListToolsResult[\"tools\"]->java.util.ArrayList[0]->io.modelcontextprotocol.spec.McpSchema$Tool[\"inputSchema\"]->io.modelcontextprotocol.spec.McpSchema$JsonSchema[\"additionalProperties\"])\n\tat com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4742)\n\tat com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:4681)\n\tat io.modelcontextprotocol.json.jackson2.JacksonMcpJsonMapper.convertValue(JacksonMcpJsonMapper.java:75)\n\tat io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport.unmarshalFrom(HttpClientStreamableHttpTransport.java:669)\n\tat io.modelcontextprotocol.spec.McpClientSession.lambda$sendRequest$12(McpClientSession.java:271)\n\tat reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:113)\n\tat reactor.core.publisher.SerializedSubscriber.onNext(SerializedSubscriber.java:99)\n\tat reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.onNext(FluxTimeout.java:181)\n\tat reactor.core.publisher.MonoCreate$DefaultMonoSink.success(MonoCreate.java:176)\n\tat io.modelcontextprotocol.spec.McpClientSession.handle(McpClientSession.java:142)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:196)\n\tat reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2570)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.request(FluxPeekFuseable.java:144)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapInner.onSubscribe(FluxFlatMap.java:968)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onSubscribe(FluxPeekFuseable.java:178)\n\tat reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)\n\tat reactor.core.publisher.Mono.subscribe(Mono.java:4576)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:430)\n\tat reactor.core.publisher.SerializedSubscriber.onNext(SerializedSubscriber.java:99)\n\tat reactor.core.publisher.FluxRetryWhen$RetryWhenMainSubscriber.onNext(FluxRetryWhen.java:178)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:716)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:592)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapInner.onSubscribe(FluxFlatMap.java:959)\n\tat reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber.onSubscribe(FluxFlattenIterable.java:222)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onSubscribe(FluxPeekFuseable.java:178)\n\tat reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onSubscribe(FluxPeekFuseable.java:178)\n\tat reactor.core.publisher.FluxJust.subscribe(FluxJust.java:68)\n\tat reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:68)\n\tat reactor.core.publisher.FluxDeferContextual.subscribe(FluxDeferContextual.java:57)\n\tat reactor.core.publisher.Flux.subscribe(Flux.java:8891)\n\tat reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:202)\n\tat reactor.core.publisher.FluxFlatMap.subscribeOrReturn(FluxFlatMap.java:94)\n\tat reactor.core.publisher.Flux.subscribe(Flux.java:8876)\n\tat reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:430)\n\tat reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onNext(MonoFlatMapMany.java:251)\n\tat reactor.core.publisher.FluxCreate$BufferAsyncSink.drain(FluxCreate.java:880)\n\tat reactor.core.publisher.FluxCreate$BufferAsyncSink.next(FluxCreate.java:805)\n\tat reactor.core.publisher.FluxCreate$SerializedFluxSink.next(FluxCreate.java:163)\n\tat io.modelcontextprotocol.client.transport.ResponseSubscribers$SseLineSubscriber.hookOnNext(ResponseSubscribers.java:150)\n\tat io.modelcontextprotocol.client.transport.ResponseSubscribers$SseLineSubscriber.hookOnNext(ResponseSubscribers.java:72)\n\tat reactor.core.publisher.BaseSubscriber.onNext(BaseSubscriber.java:160)\n\tat org.reactivestreams.FlowAdapters$FlowToReactiveSubscriber.onNext(FlowAdapters.java:211)\n\tat java.net.http/jdk.internal.net.http.LineSubscriberAdapter$LineSubscription.loop(LineSubscriberAdapter.java:447)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:182)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:207)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:280)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:233)\n\tat java.net.http/jdk.internal.net.http.LineSubscriberAdapter$LineSubscription.submit(LineSubscriberAdapter.java:188)\n\tat java.net.http/jdk.internal.net.http.LineSubscriberAdapter.onNext(LineSubscriberAdapter.java:95)\n\tat java.net.http/jdk.internal.net.http.LineSubscriberAdapter.onNext(LineSubscriberAdapter.java:53)\n\tat java.net.http/jdk.internal.net.http.common.HttpBodySubscriberWrapper.onNext(HttpBodySubscriberWrapper.java:391)\n\tat java.net.http/jdk.internal.net.http.common.HttpBodySubscriberWrapper.onNext(HttpBodySubscriberWrapper.java:49)\n\tat java.net.http/jdk.internal.net.http.ResponseContent$ChunkedBodyParser.accept(ResponseContent.java:237)\n\tat java.net.http/jdk.internal.net.http.ResponseContent$ChunkedBodyParser.accept(ResponseContent.java:131)\n\tat java.net.http/jdk.internal.net.http.Http1Response$BodyReader.handle(Http1Response.java:707)\n\tat java.net.http/jdk.internal.net.http.Http1Response$BodyReader.handle(Http1Response.java:635)\n\tat java.net.http/jdk.internal.net.http.Http1Response$Receiver.accept(Http1Response.java:527)\n\tat java.net.http/jdk.internal.net.http.Http1Response$BodyReader.tryAsyncReceive(Http1Response.java:665)\n\tat java.net.http/jdk.internal.net.http.Http1AsyncReceiver.flush(Http1AsyncReceiver.java:233)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:182)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:207)\n\tat java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.execute(HttpClientImpl.java:177)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:282)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:251)\n\tat java.net.http/jdk.internal.net.http.Http1AsyncReceiver.asyncReceive(Http1AsyncReceiver.java:468)\n\tat java.net.http/jdk.internal.net.http.Http1AsyncReceiver$Http1TubeSubscriber.onNext(Http1AsyncReceiver.java:589)\n\tat java.net.http/jdk.internal.net.http.Http1AsyncReceiver$Http1TubeSubscriber.onNext(Http1AsyncReceiver.java:546)\n\tat java.net.http/jdk.internal.net.http.common.SSLTube$DelegateWrapper.onNext(SSLTube.java:210)\n\tat java.net.http/jdk.internal.net.http.common.SSLTube$SSLSubscriberWrapper.onNext(SSLTube.java:492)\n\tat java.net.http/jdk.internal.net.http.common.SSLTube$SSLSubscriberWrapper.onNext(SSLTube.java:295)\n\tat java.net.http/jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run1(SubscriberWrapper.java:316)\n\tat java.net.http/jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run(SubscriberWrapper.java:259)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:182)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:207)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:280)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:233)\n\tat java.net.http/jdk.internal.net.http.common.SubscriberWrapper.outgoing(SubscriberWrapper.java:232)\n\tat java.net.http/jdk.internal.net.http.common.SubscriberWrapper.outgoing(SubscriberWrapper.java:198)\n\tat java.net.http/jdk.internal.net.http.common.SSLFlowDelegate$Reader.processData(SSLFlowDelegate.java:470)\n\tat java.net.http/jdk.internal.net.http.common.SSLFlowDelegate$Reader$ReaderDownstreamPusher.run(SSLFlowDelegate.java:283)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:182)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)\n\tat java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:207)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.base/java.lang.Thread.run(Thread.java:1583)\n\tSuppressed: java.lang.Exception: #block terminated with an error\n\t\tat reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:104)\n\t\tat reactor.core.publisher.Mono.block(Mono.java:1779)\n\t\tat io.modelcontextprotocol.client.McpSyncClient.listTools(McpSyncClient.java:259)\n\t\tat com.epam.aidial.core.server.controller.ToolSetToolsController.fetchTools(ToolSetToolsController.java:169)\n\t\tat com.epam.aidial.core.server.controller.ToolSetToolsController.lambda$handle$1(ToolSetToolsController.java:106)\n\t\tat com.epam.aidial.core.server.vertx.AsyncTaskExecutor.lambda$submit$0(AsyncTaskExecutor.java:54)\n\t\tat io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:240)\n\t\tat com.epam.aidial.core.server.vertx.AsyncTaskExecutor.lambda$submit$1(AsyncTaskExecutor.java:52)\n\t\tat java.base/java.util.concurrent.ThreadPerTaskExecutor$TaskRunner.run(ThreadPerTaskExecutor.java:314)\n\t\tat java.base/java.lang.VirtualThread.run(VirtualThread.java:329)\nCaused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `java.lang.Boolean` from Object value (token `JsonToken.START_OBJECT`)\n at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: io.modelcontextprotocol.spec.McpSchema$ListToolsResult[\"tools\"]->java.util.ArrayList[0]->io.modelcontextprotocol.spec.McpSchema$Tool[\"inputSchema\"]->io.modelcontextprotocol.spec.McpSchema$JsonSchema[\"additionalProperties\"])\n\tat com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:72)\n\tat com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1814)\n\tat com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1588)\n\tat com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1493)\n\tat com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseBoolean(StdDeserializer.java:533)\n\tat com.fasterxml.jackson.databind.deser.std.NumberDeserializers$BooleanDeserializer.deserialize(NumberDeserializers.java:227)\n\tat com.fasterxml.jackson.databind.deser.std.NumberDeserializers$BooleanDeserializer.deserialize(NumberDeserializers.java:200)\n\tat com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:562)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:603)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:446)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1499)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:340)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)\n\tat com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:562)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:603)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:446)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1499)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:340)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)\n\tat com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeNoNullChecks(CollectionDeserializer.java:545)\n\tat com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:358)\n\tat com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)\n\tat com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:29)\n\tat com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:562)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:603)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:446)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1499)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:340)\n\tat com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)\n\tat com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4737)\n\t... 88 more\n"},"TraceId":"494bb4aba2398fbe77ac9f7c9cc00ebb","SpanId":"c38d34d7dd01ea9e","TraceFlags":"01","Timestamp":1779353520736000000,"ObservedTimestamp":1779353520736000000,"SeverityText":"ERROR","SeverityNumber":17,"Resource":{"service.name":"aidial-core","service.version":"0.44.0-dev.15"}}
{"Body":"Responding with error. Body: Failed to fetch tools from MCP server","Attributes":{"instant":"2026-05-21T08:52:00.737860736Z","loggerName":"com.epam.aidial.core.server.ProxyContext","threadName":"vert.x-eventloop-thread-1","user.project":null,"user.id":"26fa5438-5503-41a7-af7b-2dd7ad794c2c","request.method":"GET","request.uri":"/v1/toolset/toolsets/public/27.03%20notion%20toolset__0.0.3/tools","response.status":"Bad Gateway","response.status.code":502},"TraceId":"494bb4aba2398fbe77ac9f7c9cc00ebb","SpanId":"c38d34d7dd01ea9e","TraceFlags":"01","Timestamp":1779353520737000000,"ObservedTimestamp":1779353520737000000,"SeverityText":"WARN","SeverityNumber":13,"Resource":{"service.name":"aidial-core","service.version":"0.44.0-dev.15"}}

```

### Additional information
Response from notion
<img width="398" height="155" alt="Image" src="https://github.com/user-attachments/assets/6b37fdac-2050-4b82-b66e-fc0afc0acae2" />

Mcp lib object
<img width="855" height="218" alt="Image" src="https://github.com/user-attachments/assets/1615257d-dca8-47fa-a8d2-2972b323302e" />

_No response_

--- a/server/src/main/java/com/epam/aidial/core/server/controller/ToolSetToolsController.java
+++ b/server/src/main/java/com/epam/aidial/core/server/controller/ToolSetToolsController.java
@@ -31,18 +31,27 @@ import com.epam.aidial.core.storage.http.HttpStatus;
 import com.epam.aidial.core.storage.resource.ResourceDescriptor;
 import com.epam.aidial.core.storage.resource.ResourceTypes;
 import com.epam.aidial.core.storage.util.UrlUtil;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.deser.DeserializationProblemHandler;
 import com.fasterxml.jackson.databind.node.ArrayNode;
 import com.fasterxml.jackson.databind.node.ObjectNode;
 import io.modelcontextprotocol.client.McpClient;
 import io.modelcontextprotocol.client.McpSyncClient;
 import io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport;
 import io.modelcontextprotocol.client.transport.McpHttpClientTransportAuthorizationException;
+import io.modelcontextprotocol.json.McpJsonMapper;
+import io.modelcontextprotocol.json.jackson2.JacksonMcpJsonMapper;
 import io.modelcontextprotocol.json.schema.JsonSchemaValidator;
 import io.modelcontextprotocol.spec.McpSchema;
 import io.vertx.core.Future;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.lang3.exception.ExceptionUtils;
 
+import java.io.IOException;
 import java.net.http.HttpRequest;
 import java.time.Duration;
 import java.util.List;
@@ -55,6 +64,8 @@ public class ToolSetToolsController implements Controller {
     private static final JsonSchemaValidator NOOP_SCHEMA_VALIDATOR =
             (Map<String, Object> schema, Object content) ->
                     JsonSchemaValidator.ValidationResponse.asValid(null);
+
+    private static final McpJsonMapper MCP_JSON_MAPPER = createLenientMcpJsonMapper();
 
     private final String toolSetId;
     private final boolean filterAllowed;
@@ -152,6 +163,7 @@ public class ToolSetToolsController implements Controller {
 
         HttpClientStreamableHttpTransport transport = HttpClientStreamableHttpTransport
                 .builder(upstream.getEndpoint())
+                .jsonMapper(MCP_JSON_MAPPER)
                 .httpRequestCustomizer((builder, method, endpoint, body, transportContext) ->
                         customizeRequest(builder, deployment))
                 .build();
@@ -256,6 +268,26 @@ public class ToolSetToolsController implements Controller {
         ApiKeyData.initFromContext(proxyApiKeyData, context);
         apiKeyStore.assignPerRequestApiKey(proxyApiKeyData);
         return proxyApiKeyData.getPerRequestKey();
+    }
+
+    private static McpJsonMapper createLenientMcpJsonMapper() {
+        ObjectMapper mapper = new ObjectMapper();
+        // JSON Schema allows `additionalProperties` to be a boolean OR a schema object,
+        // but the MCP SDK models it as Boolean. Coerce object/array values to null so a
+        // single non-conforming field does not fail the whole tools/list response (#1561).
+        mapper.addHandler(new DeserializationProblemHandler() {
+            @Override
+            public Object handleUnexpectedToken(DeserializationContext ctxt, JavaType targetType,
+                    JsonToken t, JsonParser p, String failureMsg) throws IOException {
+                if (targetType.hasRawClass(Boolean.class)
+                        && (t == JsonToken.START_OBJECT || t == JsonToken.START_ARRAY)) {
+                    p.skipChildren();
+                    return null;
+                }
+                return super.handleUnexpectedToken(ctxt, targetType, t, p, failureMsg);
+            }
+        });
+        return new JacksonMcpJsonMapper(mapper);
     }
 
     private void handleError(Throwable error) {

--- a/server/src/main/java/com/epam/aidial/core/server/controller/ToolSetToolsController.java
+++ b/server/src/main/java/com/epam/aidial/core/server/controller/ToolSetToolsController.java
@@ -270,6 +270,8 @@ public class ToolSetToolsController implements Controller {
         return proxyApiKeyData.getPerRequestKey();
     }
 
+    private static final String ADDITIONAL_PROPERTIES_FIELD = "additionalProperties";
+
     private static McpJsonMapper createLenientMcpJsonMapper() {
         ObjectMapper mapper = new ObjectMapper();
         // JSON Schema allows `additionalProperties` to be a boolean OR a schema object,
@@ -280,7 +282,8 @@ public class ToolSetToolsController implements Controller {
             public Object handleUnexpectedToken(DeserializationContext ctxt, JavaType targetType,
                     JsonToken t, JsonParser p, String failureMsg) throws IOException {
                 if (targetType.hasRawClass(Boolean.class)
-                        && (t == JsonToken.START_OBJECT || t == JsonToken.START_ARRAY)) {
+                        && (t == JsonToken.START_OBJECT || t == JsonToken.START_ARRAY)
+                        && ADDITIONAL_PROPERTIES_FIELD.equals(p.currentName())) {
                     p.skipChildren();
                     return null;
                 }

--- a/server/src/test/java/com/epam/aidial/core/server/ToolSetApiTest.java
+++ b/server/src/test/java/com/epam/aidial/core/server/ToolSetApiTest.java
@@ -2289,6 +2289,45 @@ public class ToolSetApiTest extends ResourceBaseTest {
     }
 
     @Test
+    void testGetAllTools_ObjectAdditionalProperties() throws JsonProcessingException {
+        // JSON Schema allows `additionalProperties` to be a schema object (e.g. Notion MCP server);
+        // the MCP SDK models it as Boolean. Such tools must not fail the whole tools/list (#1561).
+        String mcpToolsListResponse = """
+                {
+                   "jsonrpc": "2.0",
+                   "id": 2,
+                   "result": {
+                     "tools": [
+                       {"name": "search", "description": "Search",
+                        "inputSchema": {"type": "object",
+                          "properties": {"q": {"type": "string"}},
+                          "additionalProperties": {"type": "string"}}},
+                       {"name": "lookup", "description": "Lookup",
+                        "inputSchema": {"type": "object",
+                          "properties": {"id": {"type": "string"}},
+                          "additionalProperties": true}}
+                     ]
+                   }
+                 }
+                """;
+        try (TestWebServer ignore = new TestWebServer(9876, mcpToolsHandler(mcpToolsListResponse))) {
+            Response resp = send(HttpMethod.GET, "/v1/toolset/git/tools",
+                    null, null, "authorization", "admin");
+
+            assertEquals(200, resp.status());
+            var json = ProxyUtil.MAPPER.readTree(resp.body());
+            ArrayNode tools = Optional.ofNullable(json.get("tools"))
+                    .map(node -> (ArrayNode) node)
+                    .orElse(ProxyUtil.MAPPER.createArrayNode());
+            assertEquals(2, tools.size());
+            assertEquals("search", tools.get(0).get("name").asText());
+            assertEquals("lookup", tools.get(1).get("name").asText());
+            // object-valued additionalProperties is coerced to null (omitted); boolean is preserved
+            assertTrue(tools.get(1).get("inputSchema").get("additionalProperties").asBoolean());
+        }
+    }
+
+    @Test
     void testGetAllTools_RegularUserForbidden() {
         try (TestWebServer ignore = new TestWebServer(9876, mcpToolsHandler("{}"))) {
             Response resp = send(HttpMethod.GET, "/v1/toolset/git/tools");

